src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,397 @@
+                if (StatementClosingDelimiterIsMissing(currentNode))
+                {
+                    // Example: missing final `)` in `do { } while (x$$`
+                    return;
+                }
+
+            var semicolonPosition = GetSemicolonLocation(currentNode, caretPosition);
+
+            // Place cursor after the statement
+            args.TextView.TryMoveCaretToAndEnsureVisible(args.SubjectBuffer.CurrentSnapshot.GetPoint(GetEndPosition(root, semicolonPosition, currentNode.Kind())));
+        }
+
+        private int GetSemicolonLocation(SyntaxNode currentNode, int caretPosition)
+        {
+            if (currentNode.IsKind(SyntaxKind.ForStatement))
+            {
+                // in for statements, semicolon can go after initializer or after condition, depending on where the caret is located
+                var forStatementSyntax = (ForStatementSyntax)currentNode;
+                if (caretPosition > forStatementSyntax.Condition.SpanStart  && caretPosition < forStatementSyntax.Condition.Span.End)
+                {
+                    return forStatementSyntax.Condition.FullSpan.End;
+                }
+
+                if (caretPosition > forStatementSyntax.Declaration.Span.Start && caretPosition < forStatementSyntax.Declaration.Span.End)
+                {
+                    return forStatementSyntax.Declaration.FullSpan.End;
+                }
+
+                return forStatementSyntax.Incrementors.FullSpan.End;
+            }
+
+            return currentNode.Span.End;
+        }
+
+        /// <summary>
+        /// Examines the enclosing statement-like syntax for an expression which is eligible for statement completion.
+        /// </summary>
+        /// <remarks>
+        /// <para>This method tries to identify <paramref name="currentNode"/> as a node located within an argument
+        /// list, where the immediately-containing statement resembles an "expression statement". This method returns
+        /// <see langword="true"/> if the node matches a recognizable pattern of this form.</para>
+        /// </remarks>
+        private static bool LooksLikeNodeInArgumentListForStatementCompletion(SyntaxNode currentNode, 
+            ISyntaxFactsService syntaxFacts, bool isCaretAtEndOfLine)
+        {
+            // work our way up the tree, looking for a node of interest within the current statement
+            bool nodeFound = false;
+            while (!IsStatementOrFieldDeclaration(currentNode, syntaxFacts))
+            {
+                if (currentNode.IsKind(SyntaxKind.ArgumentList, SyntaxKind.ArrayRankSpecifier, SyntaxKind.ParenthesizedExpression))
+                {
+                    // It's a node of interest
+                    nodeFound = true;
+                }
+
+                // No special action is performed at this time if `;` is typed inside a string, including
+                // interpolated strings.  
+                if (IsInAString(currentNode, isCaretAtEndOfLine))
```

Comment:
What about thins like inactive regions?  NOte: we hae helpers to say "is this in a place that isn't really code".  We should likely use thather.

Reply:
Since we aren't making any text buffer edits ourselves, I don't think we need any special handling for inactive regions. The worst that will happen is typing <kbd>;</kbd> will move the caret without changing any text. Perhaps a little weird, but not horrible.

Reply:
my overall point is that we have helpers for this already.  i.e. IsInNonUserCode.  Can we use that instead of having a one off that basically does the same?

Reply:
In addition to ensuring a text-literal `;` is inserted correctly, the purpose of string handling here is to ensure cases like the following don't move the caret to an undesired location for semicolon insertion:

```csharp
var temp = $"{obj.ToString($$)}"
```

A test for semicolon in this scenario should be added. Eventually the caret could be moved outside the interpolated string, but for now the semicolon is simply added where the user types it.